### PR TITLE
Automated hyperparameter selection framework with optuna

### PR DESCRIPTION
**Pull request by qzheng75**
# Summary for changes implemented:

## Under `wod_predictor/hypereparam_search`

- `tuner.py`: the main class `HyperparamTuner` and the helper dataclass `ParamRange` for using optuna to tune the hyperparameters and defining a tunable parameter
- `objective.py`: currently only implement one objective that tunes only hyperparameters for the model with MAE as the metric
- `helpers.py`: several helper methods

## Under `modeling/kenneth`

- `configs/random_forest_model_only.yml`: the configuration file used to perform hyperparameter search on random forest
- `example_hyperparam_pipeline.ipynb`: notebook for running the hyperparameter search experiment
- `model_only_script.py`: the script version for the above notebook. Run with `python model_only_script.py --config_path=configs/random_forest_model_only.yml`

## Two bug fix in the regular pipeline

- Deep copy the input configs when instantiating the scaler so that the passed-in dictionary is not modified in-place.

### CODE REVIEW

File: wod_predictor/modeling.py [hunk 1]
```diff
@@ -127,18 +130,20 @@ def fit(self, X, y, epochs=1000, lr=0.001):
             optimizer.step()
 
             # Print every 100 epochs
-            if epoch % 100 == 0:
+            if epoch % 100 == 0 and self.verbose:
```
Comment (qzheng75) [inline wod_predictor/modeling.py:133]: Add a verbose option here

**Review by hassannaveed1997 — APPROVED** [review]
Just gave it a run, worked great! I was able to tune a Neural network and didn't cause errors. Approved!

BTW, if you can move it to the examples folder under hyper-parameter tuning that would help keep it for general use people can come across.

File: wod_predictor/hyperparam_search/helpers.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        objects= ['open_results','descriptions','benchmark_stats', 'athlete_info']
+    )
+    data = loader.load()
+
+    val_test_ratio = val_ratio + test_ratio
+    splitter = DataSplitter(
+        sample=sample, test_ratio=val_test_ratio, test_filter='23.*'
+    )
+    val_test_splitter = DataSplitter(
+        sample=int(sample * val_test_ratio),
```
[the remark this answers is not in the document]
Comment (qzheng75) [reply wod_predictor/hyperparam_search/helpers.py:54]: This makes sense. The val_test_splitter should use all samples in val_test set. Will update this with an additional commit.
